File: pages/api/decision_engine.ts
// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
//@ts-nocheck
import type { NextApiRequest, NextApiResponse } from "next";

import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== "POST") {
    return res.status(405).end();
  }
  let errorMsg = "";
  try {
    const { preAssessment, company, provider, year, profitOrLoss, loanAmount } =
      req.body;
    if (!preAssessment || typeof preAssessment !== "number") {
      errorMsg = "Invalid preAssessment";
      throw new Error(errorMsg);
    } else if (!company || typeof company !== "string") {
      errorMsg = "Invalid company";
      throw new Error(errorMsg);
    } else if (!year || typeof year !== "string") {
      errorMsg = "Invalid year";
      throw new Error(errorMsg);
    } else if (!profitOrLoss || typeof profitOrLoss !== "number") {
      errorMsg = "Invalid profitOrLoss";
      throw new Error(errorMsg);
    } else if (!provider || typeof provider !== "string") {
      errorMsg = "Invalid provider";
      throw new Error(errorMsg);
    } else if (!loanAmount || typeof loanAmount !== "string") {
      errorMsg = "Invalid loan amount";
      throw new Error(errorMsg);
    }

    const randomNumber = Math.random();
    const isApproved = randomNumber <= preAssessment / 100;
    const existingApplication = await prisma.application.findUnique({
      where: {
        company_applicationProvider: { company, applicationProvider: provider },
      },
    });
    if (existingApplication) {
      errorMsg = "Duplicate application";
      throw new Error(errorMsg);
    }
    const application = await prisma.application.create({
      data: {
        company,
        yearEstablished: new Date(year),
        profitOrLoss: profitOrLoss,
        loanAmount: parseInt(loanAmount),
        status: isApproved ? "Approved" : "Rejected",
        applicationProvider: provider,
      },
    });
    res.status(200).json({
      isApproved: isApproved,
      message: "Your application has been submitted.",
      application: application,
    });
  } catch (error) {
    return res.status(400).json({ message: errorMsg });
  }
}

File: backend/resources/decision_engine.py
import random
from datetime import datetime

from flask import request
from flask_restful import Resource

from extensions import db
from models import ApplicationsModel
from schemas import ApplicationsSchema


class DecisionEngineResource(Resource):
    def post(self):
        body = request.json
        if "preAssessment" not in body:
            return {"message": "Invalid preAssessment"}, 400
        if "company" not in body:
            return {"message": "Invalid company"}, 400
        if "year" not in body:
            return {"message": "Invalid year"}, 400
        if "profitOrLoss" not in body:
            return {"message": "Invalid profitOrLoss"}, 400
        if "provider" not in body:
            return {"message": "Invalid provider"}, 400
        if "loanAmount" not in body:
            return {"message": "Invalid loanAmount"}, 400
        company, year, pre_assessment, profit_or_loss, provider, loan_amount = (
            body["company"],
            body["year"],
            body["preAssessment"],
            body["profitOrLoss"],
            body["provider"],
            body["loanAmount"],
        )
        random_num = random.random()
        is_approved = random_num <= pre_assessment / 100
        status = "Approved" if is_approved else "Rejected"

        existing = ApplicationsModel.query.filter_by(
            company=company, applicationProvider=provider
        ).first()
        if existing:
            return {"message": "Duplicate application"}, 400
        schema = ApplicationsSchema()
        new_application = ApplicationsModel(
            company=company,
            yearEstablished=datetime.strptime(year, "%Y-%m"),
            applicationProvider=provider,
            loanAmount=loan_amount,
            status=status,
            profitOrLoss=profit_or_loss,
        )
        db.session.add(new_application)
        db.session.commit()
        return {
            "message": "Your application has been submitted.",
            "application": schema.dump(new_application),
        }, 200
